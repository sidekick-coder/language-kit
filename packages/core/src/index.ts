export * from './BaseNode'

export * from './BaseProcessor'

export * from './BaseParser'

export * from './NodeArray'
