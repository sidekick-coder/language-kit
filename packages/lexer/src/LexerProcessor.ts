import type { Token } from './Token'

export interface LexerProcessor {
    order: number
    process: (char: string, chars: string[], tokens: Token[]) => boolean
}
