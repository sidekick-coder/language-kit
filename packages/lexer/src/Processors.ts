import BreakLineProcessor from "./processors/break-line";
import SymbolProcessor from "./processors/symbol";
import WhiteSpaceProcessor from "./processors/white-space";
import WordProcessor from "./processors/word";

export const breakLineProcessor = new BreakLineProcessor()
export const symbolProcessor = new SymbolProcessor()
export const whiteSpaceProcessor = new WhiteSpaceProcessor()
export const wordProcessor = new WordProcessor()

export const allProcessors = [
    breakLineProcessor,
    symbolProcessor,
    whiteSpaceProcessor,
    wordProcessor
]