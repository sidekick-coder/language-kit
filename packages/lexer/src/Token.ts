export enum TokenType {
    Word = 'Word',
    Symbol = 'Symbol',
    Hashtag = 'Hashtag',
    WhiteSpace = 'WhiteSpace',
    BreakLine = 'BreakLine',
    EndOfFile = 'EndOfFile',
}

// TODO: add startIndex and endIndex
export class Token {
    public type: string
    public value: string
    public start = 0
    public end = 0

    public static types = TokenType

    public static from<T extends Token>(this: new (args: T) => T, type: string, value: string) {
        const token = new this({} as T)

        token.type = type
        token.value = value

        return token
    }

    public static symbol<T extends Token>(this: new (args: T) => T, value: string) {
        const token = new this({} as T)

        token.type = TokenType.Symbol
        token.value = value

        return token
    }

    public static word<T extends Token>(this: new (args: T) => T, value: string) {
        const token = new this({} as T)

        token.type = TokenType.Word
        token.value = value

        return token
    }

    public static whiteSpace<T extends Token>(this: new (args: T) => T, value = ' ') {
        const token = new this({} as T)

        token.type = TokenType.WhiteSpace
        token.value = value

        return token
    }

    public static breakLine<T extends Token>(this: new (args: T) => T, value = '\n') {
        const token = new this({} as T)

        token.type = TokenType.BreakLine
        token.value = value

        return token
    }

    public static endOfFile<T extends Token>(this: new (args: T) => T) {
        const token = new this({} as T)

        token.type = TokenType.EndOfFile
        token.value = ''

        return token
    }
}
