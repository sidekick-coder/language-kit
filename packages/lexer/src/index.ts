export * from './Processors'

export * from './Token'

export * from './Lexer'