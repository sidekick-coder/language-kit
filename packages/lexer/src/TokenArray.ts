import { Token, TokenType } from './Token'

export class TokenArray<T extends Token> extends Array<T> {
    public setPositions() {
        let position = 0

        this.forEach((token) => {
            if (token.type === TokenType.EndOfFile) {
                token.start = -1
                token.end = -1
                return
            }

            token.start = position
            position += token.value.length
            token.end = position - 1
        })
    }

    public slice(start?: number, end?: number) {
        return new TokenArray(...super.slice(start, end))
    }

    public filter<S extends T>(
        predicate: (value: T, index: number, array: T[]) => value is S,
        thisArg?: any
    ): TokenArray<S>
    public filter(
        predicate: (value: T, index: number, array: T[]) => unknown,
        thisArg?: any
    ): TokenArray<T>
    public filter(predicate: unknown, thisArg?: unknown): TokenArray<T> {
        return new TokenArray(...super.filter(predicate as any, thisArg))
    }
}
