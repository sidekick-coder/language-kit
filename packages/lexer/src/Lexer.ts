import type { LexerProcessor } from './LexerProcessor'
import { Token } from './Token'

import { allProcessors } from './Processors'
import { TokenArray } from './TokenArray'

export interface LexerTokenizeOptions {
    includeEndOfFileToken?: boolean
}

const defaultOptions: LexerTokenizeOptions = {
    includeEndOfFileToken: true,
}

export class Lexer<T extends Token = Token> {
    public processors: LexerProcessor[] = allProcessors

    /**
     * Tokenize a string of code
     * @param code
     * @param options
     * @returns A TokenArray
     * @example
     * ```ts
     * const lexer = new Lexer()
     * const tokens = lexer.tokenize('1 + 1')
     * ```
     */
    public tokenize(code: string, options: LexerTokenizeOptions = defaultOptions) {
        const tokens = new TokenArray<T>()

        const chars = code.split('')

        while (chars.length) {
            const current = chars[0]

            this.processors.sort((a, b) => a.order - b.order)

            const result = this.processors.find((p) => p.process(current, chars, tokens))

            if (result) continue

            console.debug('[@language-kit/lexer] unhandled char', current)

            chars.shift()
        }

        if (options.includeEndOfFileToken) {
            tokens.push(Token.endOfFile() as T)
        }

        tokens.setPositions()

        return tokens
    }
}
