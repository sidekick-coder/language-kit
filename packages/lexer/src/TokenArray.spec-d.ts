import { describe, it, expectTypeOf } from 'vitest'
import { Token } from '.'
import { TokenArray } from './TokenArray'

describe('TokenArray.d.ts', () => {
    const methods = ['setPositions'] as const

    it('should slice() return an instance of TokenArray', () => {
        const tokens = new TokenArray()

        tokens.push(...[Token.word('Hello'), Token.whiteSpace(), Token.word('word')])

        const result = tokens.slice(0, 1)

        methods.forEach((method) => {
            expectTypeOf(result).toEqualTypeOf<InstanceType<typeof TokenArray>>()

            expectTypeOf(result[method]).toBeFunction()
        })
    })

    it('should map() return an instance of TokenArray', () => {
        const tokens = new TokenArray()

        tokens.push(...[Token.word('Hello'), Token.whiteSpace(), Token.word('word')])

        const result = tokens.map((t) => t)

        expectTypeOf(result).toEqualTypeOf<InstanceType<typeof Array<Token>>>()
    })

    it('should filter() return an instance of TokenArray', () => {
        const tokens = new TokenArray()

        tokens.push(...[Token.word('Hello'), Token.whiteSpace(), Token.word('word')])

        const result = tokens.filter((t) => t)

        expectTypeOf(result).toEqualTypeOf<InstanceType<typeof TokenArray>>()
    })
})
