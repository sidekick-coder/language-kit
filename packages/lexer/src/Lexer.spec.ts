import { describe, it, expect } from 'vitest'
import { Lexer } from '../src/Lexer'
import { Token } from '.'

describe('lexer', () => {
  const lexer = new Lexer()

  const files = import.meta.glob('./tests/fixtures/*.txt', {
    eager: true,
    as: 'raw'
  })

  it.each(Object.entries(files))('should tokenize fixtures %s', (filename, content) => {
    const tokens = lexer.tokenize(content)

    const length = tokens.reduce((acc, token) => acc + token.value.length, 0)

    expect(lexer.tokenize(content)).toMatchSnapshot()

    expect(length).toBe(content.length)
  })

  it('should correctly convert whitespaces', () => {
    const tokens = lexer.tokenize('This is a **bold text** \n\n')

    const expected = [
      Token.word('This'),
      Token.whiteSpace(),
      Token.word('is'),
      Token.whiteSpace(),
      Token.word('a'),
      Token.whiteSpace(),
      Token.symbol('*'),
      Token.symbol('*'),
      Token.word('bold'),
      Token.whiteSpace(),
      Token.word('text'),
      Token.symbol('*'),
      Token.symbol('*'),
      Token.whiteSpace(),
      Token.breakLine(),
      Token.breakLine(),
      Token.endOfFile()
    ]

    expect(tokens).toEqual(expected)
  })
})
