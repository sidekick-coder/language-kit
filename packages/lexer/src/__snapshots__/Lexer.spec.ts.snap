// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lexer > should tokenize fixtures ./tests/fixtures/symbols.txt 1`] = `
[
  Token {
    "type": "Symbol",
    "value": "!",
  },
  Token {
    "type": "Symbol",
    "value": "@",
  },
  Token {
    "type": "Symbol",
    "value": "#",
  },
  Token {
    "type": "Symbol",
    "value": "$",
  },
  Token {
    "type": "Symbol",
    "value": "%",
  },
  Token {
    "type": "Symbol",
    "value": "^",
  },
  Token {
    "type": "Symbol",
    "value": "&",
  },
  Token {
    "type": "Symbol",
    "value": "*",
  },
  Token {
    "type": "Symbol",
    "value": "(",
  },
  Token {
    "type": "Symbol",
    "value": ")",
  },
  Token {
    "type": "Symbol",
    "value": "_",
  },
  Token {
    "type": "Symbol",
    "value": "+",
  },
  Token {
    "type": "Symbol",
    "value": "-",
  },
  Token {
    "type": "Symbol",
    "value": "=",
  },
  Token {
    "type": "Symbol",
    "value": "[",
  },
  Token {
    "type": "Symbol",
    "value": "]",
  },
  Token {
    "type": "Symbol",
    "value": "{",
  },
  Token {
    "type": "Symbol",
    "value": "}",
  },
  Token {
    "type": "Symbol",
    "value": ";",
  },
  Token {
    "type": "Symbol",
    "value": "'",
  },
  Token {
    "type": "Symbol",
    "value": ":",
  },
  Token {
    "type": "Symbol",
    "value": "\\"",
  },
  Token {
    "type": "Symbol",
    "value": ",",
  },
  Token {
    "type": "Symbol",
    "value": ".",
  },
  Token {
    "type": "Symbol",
    "value": "/",
  },
  Token {
    "type": "Symbol",
    "value": "<",
  },
  Token {
    "type": "Symbol",
    "value": ">",
  },
  Token {
    "type": "Symbol",
    "value": "?",
  },
  Token {
    "type": "EndOfFile",
    "value": "",
  },
]
`;

exports[`lexer > should tokenize fixtures ./tests/fixtures/text.txt 1`] = `
[
  Token {
    "type": "Word",
    "value": "Lorem",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "ipsum",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "dolor",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "sit",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "amet",
  },
  Token {
    "type": "Symbol",
    "value": ",",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "consectetur",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "adipiscing",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "elit",
  },
  Token {
    "type": "Symbol",
    "value": ".",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "Word",
    "value": "Fusce",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "varius",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "vestibulum",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "velit",
  },
  Token {
    "type": "Symbol",
    "value": ",",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "ut",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "lobortis",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "massa",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "sagittis",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "at",
  },
  Token {
    "type": "Symbol",
    "value": ".",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "Word",
    "value": "Orci",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "varius",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "natoque",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "penatibus",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "et",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "magnis",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "dis",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "parturient",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "montes",
  },
  Token {
    "type": "Symbol",
    "value": ",",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "nascetur",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "ridiculus",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "mus",
  },
  Token {
    "type": "Symbol",
    "value": ".",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "Nullam",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "imperdiet",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "tincidunt",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "posuere",
  },
  Token {
    "type": "Symbol",
    "value": ".",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "Word",
    "value": "Aliquam",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "et",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "neque",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "a",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "augue",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "iaculis",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "luctus",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "rutrum",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "vel",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "ipsum",
  },
  Token {
    "type": "Symbol",
    "value": ".",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "Word",
    "value": "Integer",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "aliquam",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "nibh",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "non",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "scelerisque",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "viverra",
  },
  Token {
    "type": "Symbol",
    "value": ".",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "Word",
    "value": "Cras",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "at",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "semper",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "urna",
  },
  Token {
    "type": "Symbol",
    "value": ".",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "Word",
    "value": "Proin",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "nibh",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "odio",
  },
  Token {
    "type": "Symbol",
    "value": ",",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "feugiat",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "ullamcorper",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "est",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "vitae",
  },
  Token {
    "type": "Symbol",
    "value": ",",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "vestibulum",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "hendrerit",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "nulla",
  },
  Token {
    "type": "Symbol",
    "value": ".",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "Word",
    "value": "Fusce",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "lacinia",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "sapien",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "at",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "lorem",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "mattis",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "fringilla",
  },
  Token {
    "type": "Symbol",
    "value": ".",
  },
  Token {
    "type": "EndOfFile",
    "value": "",
  },
]
`;
